[docs] add localization doc (#462)
microsoft/vcpkg-tool: docs/localization.md

# Writing Localizeable Code

As part of Microsoft's commitment to global-ready products, our developer tools are localized into 14 different languages.
For this effort, vcpkg has hooks in order to allow our localization team to translate all messages that are shown to the user by the vcpkg tool.
As a developer on vcpkg, the only thing you have to consider is the messages in the C++ source files,
and `locales/messages.json` -- everything else will be generated and modified by the localization team.

## Declaring a Message

The process of writing a message to be displayed to the user starts with the message; if it needs to be in a header,

> use the `DECLARE_MESSAGE(<message-name>, <parameters>, <comment>, <english-message>` in that header,

Reviewer: ```suggestion
use the `DECLARE_MESSAGE(<message-name>, <parameters>, <comment>, <english-message>)` in that header,
```

followed by `REGISTER_MESSAGE(<message-name>)` in a source file; if it can just be in a source file,

> use `DECLARE_AND_REGISTER_MESSAGE()`.

Reviewer: I think you should work in a complete example to provide context for the subheadings. For example, a direct link to `https://github.com/microsoft/vcpkg-tool/blob/13a09ef0359e259627d46560a22a6e182730da7b/src/vcpkg/sourceparagraph.cpp#L24-L28`

which looks like this when rendered by GitHub:

https://github.com/microsoft/vcpkg-tool/blob/13a09ef0359e259627d46560a22a6e182730da7b/src/vcpkg/sourceparagraph.cpp#L24-L28

> use `DECLARE_AND_REGISTER_MESSAGE()`.

Reviewer 2: For the following part I would format it something like:

=====================

Most user messages can be declared and registered in the source file in a single step with:
```
DECLARE_AND_REGISTER_MESSAGE(<message-name>, <parameters>, <comment>, <english-message>)
```

If you need to declare a message in a header file, then you need to declare it first in the header file:
```
DECLARE_MESSAGE(<message-name>, <parameters>, <comment>, <english-message>)
```
and then register it in the source file.
```
REGISTER_MESSAGE(<message-name>)
```

### Message Names

Message names should be descriptive and unique, and must be `CamelCase`.
When referring to a message in `msg::format`, you must add `msg` to the front of it; for example,
`DECLARE_MESSAGE(MyMessage, ...)` can be referred to as `msgMyMessage`.

### Parameters

> Each placeholder in a localized format string must have a corresponding parameter in this list.
> It should look something like `(msg::arg1, msg::arg2)`.
> This allows us to make sure that we don't have format errors, where one forgets a placeholder,
> or adds an extra one.
> Each parameter in the list should be used in the message.

Reviewer 2: ```suggestion
If the message contains placeholders, each one must have a corresponding parameter.
For example, `"Installing {package}:{triplet}"` should have a list parameters `(msg::package, msg:triplet)`.
This allows us to make sure that we don't have format errors, where one forgets a placeholder,
or adds an extra one.
Each parameter in the list should be used in the message.
```

### Comment

The comment is used to give context to the translator as to what a placeholder will be replaced with.
For example, in `BuildResultSummaryHeader` (`"SUMMARY FOR {triplet}"`),
it will note that an example of `{triplet}` is `'x64-windows'`.
Most message placeholders (see `messages.h` for the full list - search for `DECLARE_MSG_ARG`) have comments associated with them already.
Only general placeholders like `{value}`, `{expected}`, and `{actual}` don't have comments associated with them.
If you use any of these placeholders, write `example of {<name>} is '<expected-value>'` for each of those placeholders,

> separated by `\n`. You can also add context if you feel it's necessary as the first line.

Reviewer 2: This part is a bit hard to follow. Maybe we should start with an example of an existing message that uses all parameters and break down each part of the declaration.

> An additional, special value is `"{Locked}"` -- this is a note to the localizers that they should not modify the message.
> However, you should not use this value, and we will remove existing uses of it over time.

Reviewer: ```suggestion
```

Plus https://github.com/microsoft/vcpkg-tool/pull/463

> However, you should not use this value, and we will remove existing uses of it over time.

Reviewer 2: Why shouldn't this be used?

Reviewer: I argued that we should only be putting things that are locale sensitive into the declared and registered messages stuff, and that truly locale-invariant things should use fmt::format directly even if that means they'll be somewhat more verbose at the call site.

### Message

Messages in vcpkg are written in American English. They should not contain:

* formatting:
  - indentation should be added with the `append_indent()` function
  - newlines should be added with the `append_newline()` function

>   - Any other interesting characters (like `- ` for lists, for example) should use `append(LocalizedString::from_raw(...))`

Reviewer 3: Perhaps this indicates a need for `.append_raw(...)`.

Reviewer: You mean like https://github.com/microsoft/vcpkg-tool/blob/13a09ef0359e259627d46560a22a6e182730da7b/include/vcpkg/base/messages.h#L37-L41 ? 🤣

Author: I added it and I still forgot about it 🤣

> * the kind of message:
>   - warnings should be printed as `msg::format(msg::msgWarningMessage).append(msgMyWarning)`
>   - errors the same, with `msg::msgErrorMessage`
>   - and internal errors the same, with `msg::msgInternalErrorMessage`.

Reviewer 2: ```suggestion
* or the prefixes:
  - `"warning: "`, instead use  `msg::format(msg::msgWarningMessage).append(msgMyWarning)`
  - `"error:"`, use `msg::msgErrorMessage`
  - `"internal error:"`, use `msg::msgInternalErrorMessage`.
```

However, that's as far as you need to go -- once you've done the writing of the message, and the `x-generate-default-message-map` stuff,
you can rest easy not modifying anything else.

### Putting it All Together

Let's create a vcpkg hello world command to show off how one can use the messages API.

```cxx
// commands.hello-world.cpp
#include <vcpkg/base/messages.h>

#include <vcpkg/commands.hello-world.h>
#include <vcpkg/vcpkgcmdarguments.h>
#include <vcpkg/vcpkgpaths.h>

namespace
{
    // boilerplate
    const CommandStructure HELLO_WORLD_COMMAND_STRUCTURE = {
        create_example_string("hello-world <name>"),
        0,
        1,
        {{}, {}, {}},
        nullptr,
    };

    // note that we add additional context in the comment here
    DECLARE_AND_REGISTER_MESSAGE(World, "We will say hello to 'world' if no name is given", (), "world");
    // here, `{value}` is a placeholder that doesn't have example text, so we need to give it ourselves
    DECLARE_AND_REGISTER_MESSAGE(Hello, "example for {value} is 'world'", (msg::value), "Hello, {value}!");
    // here, `{triplet}` _already has_ example text, so it's fine to not give a comment

>     DECLARE_AND_REGISTER_MESSAGE(MyTripletIs, "", (msg::triplet), "My triplet is {triplet}.");

Reviewer: ```suggestion
    DECLARE_AND_REGISTER_MESSAGE(MyTripletIs, (msg::triplet), "", "My triplet is {triplet}.");
```

This is why I prefer linking to the real example :)

As a developer on vcpkg, the only thing you have to consider is the messages in the C++ source files,
and `locales/messages.json` -- everything else will be generated and modified by the localization team.

## Declaring a Message

The process of writing a user-visible message starts with declaring it.
Most user-facing messages can be declared and registered at the same time in a source file with:

```cxx
DECLARE_AND_REGISTER_MESSAGE(<message-name>, <parameters>, <comment>, <english-message>);
```

for example, in [`sourceparagraph.cpp`].

If you need to declare a message in a header file
(for example, if a templated function uses it),
then you need to first declare it in the header:

```cxx
DECLARE_MESSAGE(<message-name>, <parameters>, <comment>, <english-message>);
```

and then register it in the corresponding source file:

```cxx
REGISTER_MESSAGE(<message-name>);
```

An example of this lies in [`graphs.h`] and [`graphs.cpp`].

[`sourceparagraph.cpp`]: https://github.com/microsoft/vcpkg-tool/blob/13a09ef0359e259627d46560a22a6e182730da7b/src/vcpkg/sourceparagraph.cpp#L24-L28

> [`graphs.h`]: https://github.com/microsoft/vcpkg-tool/blob/13a09ef0359e259627d46560a22a6e182730da7b/include/vcpkg/graphs.h#L13

Reviewer 2: ```suggestion
[`graphs.h`]: https://github.com/microsoft/vcpkg-tool/blob/ca8099607bfa71adac301b56c601fd71d8ccab9b/include/vcpkg/base/graphs.h#L13
```

Permalink

> [`graphs.cpp`]: https://github.com/microsoft/vcpkg-tool/blob/13a09ef0359e259627d46560a22a6e182730da7b/src/vcpkg/graphs.cpp#L5

Reviewer 2: ```suggestion
[`graphs.cpp`]: https://github.com/microsoft/vcpkg-tool/blob/ca8099607bfa71adac301b56c601fd71d8ccab9b/src/vcpkg/base/graphs.cpp#L5
```

Link doesn't work for me on preview, using permalink instead.

### Message

Messages in vcpkg are written in American English. They should not contain:

* formatting:
  - indentation should be added with the `append_indent()` function;
    if you need to add more than one indentation, you can use `append_indent(N)`
  - newlines should be added with the `appendnl()` function
  - Any other interesting characters (like `- ` for lists, for example) should use `append_raw(...)`
* or for the prefixes:
  - `"warning: "`, instead use `msg::format(msg::msgWarningMessage).append(msgMyWarning)`
  - `"error: "`, instead use `msg::msgErrorMessage`
  - `"internal error: "`, instead use `msg::msgInternalErrorMessage`.

They should also not be simple, locale-invariant messages -- something like, for example,
`{file}:{line}:{column}: ` should be done with `LocalizedString::from_raw(fmt::format("{}:{}:{}", file, line, column))`.

Other than that, all messages that are printed to the user must be localizeable.
However, that's as far as you need to go -- once you've done the writing of the message, and the `x-generate-default-message-map` stuff,
you can rest easy not modifying anything else.

### Putting it All Together

> Let's create a vcpkg hello world command to show off how one can use the messages API.

Reviewer: I'm still not a fan of including a real example with a lot of code that's likely to rot rather than just linking to a real use but not over my dead body.
